### PR TITLE
LiteralCNF is_action_safe changed

### PR DESCRIPTION
**Pull request by oriKay**
First version of the cnf.is_action_safe

### CODE REVIEW

File: sam_learning/core/propositional_operations/literals_cnf.py [hunk 1]
```diff
@@ -67,6 +67,35 @@ def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bo
...
 
+    def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bool:
+        """Checks if an action is safe to execute based on this CNF clause.
+
+               :param action_name: the name of the action.
+               :param action_preconditions: the preconditions of the action.
+               :return: True if the action is safe to execute, False otherwise.
+               """
+
+        unitc = []  # unit clauses of size 1 (1 tuple) that contains action_name
```
Comment (argaman-aloni) [inline sam_learning/core/propositional_operations/literals_cnf.py:78]: unit_clauses
non_unit_clauses

Avoid abbrevations when it makes the code less readable.
If the variable names are understandable you don't need the extra docs.

File: sam_learning/core/propositional_operations/literals_cnf.py [hunk 2]
```diff
@@ -67,6 +67,35 @@ def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bo
 
         return True
 
+    def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bool:
+        """Checks if an action is safe to execute based on this CNF clause.
+
+               :param action_name: the name of the action.
+               :param action_preconditions: the preconditions of the action.
+               :return: True if the action is safe to execute, False otherwise.
+               """
```
Comment (argaman-aloni) [inline sam_learning/core/propositional_operations/literals_cnf.py:76]: de-indent

File: sam_learning/core/propositional_operations/literals_cnf.py [hunk 3]
```diff
@@ -67,6 +67,35 @@ def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bo
...
+
+        unitc = []  # unit clauses of size 1 (1 tuple) that contains action_name
+        nonunc = []  # non-unit clauses that contain action_name
+
+        for lifted_options in self.possible_lifted_effects:
+            if action_name in [action for (action, _) in lifted_options]:
+                if len(lifted_options) == 1:
+                    unitc.append(lifted_options[0])
+                else:
```
Comment (argaman-aloni) [inline sam_learning/core/propositional_operations/literals_cnf.py:85]: new line

File: sam_learning/core/propositional_operations/literals_cnf.py [hunk 4]
```diff
@@ -67,6 +67,35 @@ def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bo
...
+                    nonunc.append(lifted_options)
+
+        if len(nonunc) == 0:
+            return True
+
+        for nunc in nonunc:
+            if not any(uc in nunc for uc in unitc):
+                for (action, predicate) in nunc:
```
Comment (argaman-aloni) [inline sam_learning/core/propositional_operations/literals_cnf.py:93]: does this work?

File: sam_learning/core/propositional_operations/literals_cnf.py [hunk 5]
```diff
@@ -67,6 +67,35 @@ def is_action_safe(self, action_name: str, action_preconditions: Set[str]) -> bo
...
+                if len(lifted_options) == 1:
+                    unitc.append(lifted_options[0])
+                else:
+                    nonunc.append(lifted_options)
+
+        if len(nonunc) == 0:
+            return True
+
+        for nunc in nonunc:
+            if not any(uc in nunc for uc in unitc):
```
Comment (argaman-aloni) [inline sam_learning/core/propositional_operations/literals_cnf.py:92]: this works?

**Review by argaman-aloni — COMMENTED** [review]
Add a test based on the talk we had and see that the scenario returns what you expect it to return

File: sam_learning/core/propositional_operations/literals_cnf.py (hunk 5 above)
Comment (oriKay) [reply sam_learning/core/propositional_operations/literals_cnf.py:92]: let me better check that with some tests as you suggested

**Comment by oriKay** [comment]
this fix is inside a new and bigger branch

**Pull request closed without merging**